Replace DeadlineTimer with asio::steady_timer (RIPD-1356)

scottschurr opened the pull request:
Replaces the four remaining uses of `DeadlineTimer` with uses of `boost::asio::steady_timer`.  Since the last uses of `DeadlineTimer` are removed, that allows us to remove `DeadlineTimer` and its unit tests from the build tree.

Note that the three new places that call `JobQueue::addJob()` verify that `JobQueue::isStopped()` is false before calling `addJob()`.  During the process of testing it was discovered that without these tests the `addJob()` calls might cause an assert during shutdown.  So I've left the `isStopped()` tests in place.

Reviewers: @vinniefalco @miguelportilla

vinniefalco commented on src/ripple/app/ledger/LedgerTiming.h:
```diff
@@ -176,7 +176,7 @@ auto constexpr LEDGER_MIN_CONSENSUS = 2s;
 
 // How often we check state or change positions (in milliseconds)
-auto constexpr LEDGER_GRANULARITY = 1s;
+auto constexpr LedgerStateCheckInterval = 1s;
```
The comment says "in milliseconds"

scottschurr replied:
Good catch.  Thanks.

vinniefalco commented on src/ripple/app/main/Application.cpp:
```diff
@@ -862,39 +861,52 @@ class ApplicationImp
-            if (! config_->standalone())
+        sweepTimer_.expires_from_now (
+            std::chrono::seconds {config_->getSize (siSweepInterval)});
+        sweepTimer_.async_wait ([this] (boost::system::error_code const& e)
```
Doesn't the completion handler need to be wrapped in a strand?

vinniefalco replied:
`ec` for the name of the error code is idiomatic

scottschurr replied:
Regarding the `strand`, probably so.  You are the `asio` master.  I'll need to read up on `strand`s.

vinniefalco commented on src/ripple/app/main/Application.cpp:
```diff
@@ -862,39 +861,52 @@ class ApplicationImp
-                // VFALCO TODO Give this magic constant a name and move it into a well documented header
-                //
-                if (space.available < (512 * 1024 * 1024))
+            if ((e.value() == boost::system::errc::success) &&
```
This is idiomatic:
```
if(! ec && ...
```

vinniefalco commented on src/ripple/app/main/Application.cpp:
```diff
@@ -862,39 +861,52 @@ class ApplicationImp
+        entropyTimer_.expires_from_now (5min);
+        entropyTimer_.async_wait ([this] (boost::system::error_code const& e)
+            {
+                if (e.value() == boost::system::errc::success)
```
If there's an error we might want to log and/or terminate the app - because we won't be reseeding with fresh entropy. @nbougalis ?

nbougalis replied:
I don't think we *need* to terminate, although we could and it's probably safer if we do.

Longer term, I'd like something more elegant that does away with the need for a timer (namely, Fortuna) that can just be fed entropy directly.

vinniefalco commented on src/ripple/app/main/Application.cpp:
```diff
@@ -862,39 +861,52 @@ class ApplicationImp
+    {
+        using namespace std::chrono_literals;
+        entropyTimer_.expires_from_now (5min);
+        entropyTimer_.async_wait ([this] (boost::system::error_code const& e)
```
I think you need a strand

vinniefalco commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -474,16 +471,18 @@ class NetworkOPsImp final
+        if (clusterTimer_)
+            clusterTimer_->cancel();
 
         stopped ();
```
I believe there's a race condition here. Consider if the completion handler is already running, and we call `stopped` just before the completion handler calls `addJob`. Now we have incorrectly signaled that we have stopped, but then we continue to use resources.

scottschurr replied:
Excellent point.  What's the correct method for determining whether a completion handler is running?  I did not see a technique when I looked at `asio::steady_timer`, but I probably missed it.

vinniefalco replied:
There's no explicit way to know, you just have to structure code so that there is no race. @nbougalis ?

miguelportilla replied:
You can post to the io_service thread.

`heartbeatTimer_->get_io_service().post([=] { heartbeatTimer_->cancel(); });`

vinniefalco replied:
That also works, but is not synchronous

vinniefalco commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -526,8 +525,8 @@ class NetworkOPsImp final
-    DeadlineTimer m_heartbeatTimer;
-    DeadlineTimer m_clusterTimer;
+    boost::optional<boost::asio::steady_timer> heartbeatTimer_;
+    boost::optional<boost::asio::steady_timer> clusterTimer_;
```
This is weird, why would we put timers in an optional container?

scottschurr replied:
The `optional` is a cheaper version of `unique_ptr` that doesn't allocate.  We need to postpone the construction of the timers until after the `NetworkOPs` constructor runs because we shouldn't call `app_` until `NetworkOPs` is fully constructed.  At least that's the intention.

vinniefalco commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -619,33 +618,52 @@ NetworkOPsImp::getHostId (bool forAdmin)
+    // We cannot initialize our asio::steady_timers in the constructor
+    // since that would require a call through a virtual method of an
+    // incompletely constructed Application.  So we initialize them here,
+    // just before their first use.
```
We could just make `boost::asio::io_service&` a constructor parameter.

scottschurr replied:
Sure, that would work.  Good point.

vinniefalco commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -619,33 +618,52 @@ NetworkOPsImp::getHostId (bool forAdmin)
-    m_heartbeatTimer.setExpiration (LEDGER_GRANULARITY);
+    assert (heartbeatTimer_);
+    heartbeatTimer_->expires_from_now (LedgerStateCheckInterval);
+    heartbeatTimer_->async_wait ([this] (boost::system::error_code const& e)
```
I think you need a strand here

scottschurr commented:
I'm suspecting that at least part of the problem I'm seeing is that Stoppables are out of order after my change.  I'm now digging into the Stoppables to try and figure out what's going on.

scottschurr commented:
I'm closing this pull request (I've preserved changes addressing your comments @vinniefalco).  I'm going to take some time  to study the effects of some of the previous DeadlineTimer changes.
